Add rendering for waterway=waterfall [WIP]

meased opened the pull request:
Fixes #336

Adds icon and name rendering for waterway=waterfall.

Before

![waterfall_before_15](https://user-images.githubusercontent.com/14190496/38182551-11b2ac08-35ef-11e8-8e70-b9809b547917.png)

After

![waterfall_after_15](https://user-images.githubusercontent.com/14190496/38182556-1c6a392c-35ef-11e8-8e0d-898a2a3c4737.png)

Various zoom levels:

![waterfalls_after_17](https://user-images.githubusercontent.com/14190496/38182630-a5872670-35ef-11e8-8d94-d9b1eea2f22a.png)
![waterfall_after_14](https://user-images.githubusercontent.com/14190496/38182772-8c5f3a06-35f0-11e8-9cdf-5b50688edb60.png)
![waterfall_after_13](https://user-images.githubusercontent.com/14190496/38182792-b787a682-35f0-11e8-9517-e0f69cbd88b1.png)

This uses the waterfall v3 icon from @gpsvisualizer (I did some slight pixel alignment) with both icon and text using the standard water text color.

I'm not sure why cliffs/highways are rendering on top of the waterfall icon. Can anyone offer some advice here? Is this just kosmtik?

I'm also not too happy with the low zoom. I think waterfalls are prominant features that primarily occur in rural areas and deserve to be rendered fairly low, but I'd like to try rendering only those with names at lower zooms.

kocio-pl commented:
What about rendering height? I guess this would be the same code as in https://github.com/gravitystorm/openstreetmap-carto/pull/3158#issuecomment-378103735.

kocio-pl commented:
Please remember to add the links to the testing places, so other people can test it too.

kocio-pl commented:
I think height is an important factor to consider initial zoom level - see for example this one (just 6.1 m):

https://www.openstreetmap.org/node/4247958331#map=19/45.57022/-122.12423

This is probably the testing place:

https://www.openstreetmap.org/node/1017565710#map=16/45.5758/-122.1210

meased commented:
It may not be too hard to render height, assuming I can lift the code from natural=peak or towers. I'll give it a go.

jragusa commented:
Please note that icon is rendered below the cliff features which decreases its readability. This is well visible in picture 3 and the combination of cliff with waterfall is not so rare. The icon of the lower waterfall is even nearly unreadable.

Additional question: is it normal to separate waterfall line 2155 from line 2164 ?

meased commented:
No, that's probably not normal, I'll fix it.

As mentioned above I am aware that cliffs and highways are overlapping the icon and I suspect that it's some weirdness with kosmtik. Partly why this PR is WIP. Suggestions anyone?

jragusa commented:
@meased Oops I missed you have already mentioned this. I think that water features are localised in a lower layer than cliffs.

kocio-pl commented:
> It may not be too hard to render height, assuming I can lift the code from natural=peak or towers. I'll give it a go.

Just to be clear: I think also about setting initial zoom related to the height, just like we do with masts and towers, for example:

https://github.com/gravitystorm/openstreetmap-carto/blob/94fe461654bd1424e775a23207eed6ffb4aa2203/amenity-points.mss#L371-L375

meased commented:
The layering has been sorted out, I had added waterfalls to the water-barriers-points class, which is lower than highways (assumably to allow roads on top of dams). I moved it to the amenity-point class and everything looks good now.

I also got the height filtering working so we will be able to show taller waterfalls at lower zoom. I believe sorting by height is also working (lifted from natural=peak) so when two waterfalls overlap the taller will be preferred, though my test area only has one case of this.

I still would like to get the height rendered in the name like peaks.

I'll post some updated sample images later...

meased commented:
Height rendering is working:
![lowzoom_height](https://user-images.githubusercontent.com/14190496/38343634-ec56d2c0-3839-11e8-92af-d4bd1b0fb4cf.png)
Zoomed in shows all waterfalls:
![highzoom_height](https://user-images.githubusercontent.com/14190496/38343660-12949c42-383a-11e8-8bb8-a907f30b614e.png)
From this area: https://www.openstreetmap.org/#map=15/45.6227/-121.8954

The original area to show current icon rendering (none of these have height here, I recently added height to a handful of them for test rendering but have to reload my database to get it):
![multnomah](https://user-images.githubusercontent.com/14190496/38343732-79396fc2-383a-11e8-9a84-faae5890358a.png)

meased commented:
Here's a question: should waterfalls have an oblique font? Every other natural water feature is oblique except springs, maybe waterfalls should follow suit.
![oblique](https://user-images.githubusercontent.com/14190496/38344174-f48e726a-383c-11e8-942c-e154675bfe40.png)

gpsvisualizer commented:
That's a good question. It seems like oblique fonts are usually used on areas, whereas single points like peaks & springs tend to have non-oblique fonts.  (I'm sure there are exceptions.)  At any rate, I prefer the non-oblique font in this case.

kocio-pl commented:
I like the system you invented (height+name) and the testing you've made. I will try to make independent testing soon, but it looks really good.

gpsvisualizer commented:
Would it be possible to have smaller waterfalls appear on the map sooner (i.e., at lower zoom levels), but without their label?  That way they wouldn't clutter the map much, but there'd be an indication that something's there.

kocio-pl commented:
Thanks to @rrzefox we can now test rendering around the planet:

http://bl.ocks.org/matthijsmelissen/raw/af7a602c222dbf1ff1a2c0d84ed755b7/#13.00/45.5565/-122.1355

gpsvisualizer commented on project.mml:
```diff
@@ -1571,13 +1574,18 @@ Layer:
                 END
+              WHEN "waterway" IN ('waterfall') THEN
+                CASE
+                  WHEN tags->'height' ~ '^\d{1,4}(\.\d+)?$' THEN (tags->'height')::NUMERIC
```
Wow... I would have thought feet or feet/inches was simply not allowed!  I know that meters are *required* in the "ele" tag.

How many waterfalls are tagged with the "'" unit?  If it's not too many, maybe we should just convert them all to meters and be done with it.

meased replied:
My suggestion would be to make this PR about waterfalls and add support for feet/inches in another PR (if we decide to do it at all).

Wether or not to support an "m" unit in the string doesn't matter much to me. This handles all of the above cases except feet/inches:
``` sql
WHEN tags->'height' ~ '^\d{1,4}(\.\d+)?\s?m?$' THEN (SUBSTRING(tags->'height', '^(\d{1,4}(\.\d+)?)\s?m?$'))::NUMERIC
```

meased commented on project.mml:
```diff
@@ -2102,13 +2119,15 @@ Layer:
                 name,
-                CASE WHEN name IS NOT NULL AND elevation IS NOT NULL THEN E'\n' ELSE NULL END,
-                CASE WHEN elevation IS NOT NULL THEN CONCAT(REPLACE(ROUND(elevation)::TEXT, '-', U&'\2212'), U&'\00A0', 'm') ELSE NULL END
+                CASE WHEN elevation IS NOT NULL THEN CONCAT(E'\n', REPLACE(ROUND(elevation)::TEXT, '-', U&'\2212'), U&'\00A0', 'm') ELSE NULL END,
```
I made a mistake here.

This is the code that appends the elevation text to the name column. It does need to be modified by this PR because now both elevation and height are rendered, whereas before it was only elevation.

I tried to simplify the expression by forcing a newline before the elevation/height text, but that doesn't work when the name is null. I have put it back to the original structure.

kocio-pl replied:
I meant only this line of code, but now it's the same as before, so no problem here.

sommerluk commented:
Could `(\sm)` be replaced by `( m)` to force `U+0020` as the only valid codepoint before `m`?

meased commented:
> Could (\sm) be replaced by ( m) to force U+0020 as the only valid codepoint before m?

Seems like a good idea. Done.

gpsvisualizer commented:
Thanks, everyone, for getting this done!
